Add doc on how to use code paths for custom library (#13702)
mlflow/mlflow: docs/source/model/dependencies.rst

code like ``from utils import my_func``. You can also specify a directory path as ``code_paths`` to save multiple files under the directory:

Use of ``code_paths`` Option for a Custom Library
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To include custom libraries or libraries from a private mirror server when logging your model, use the ``code_paths`` option. 
This option allows you to upload .whl files or other dependencies alongside your model, ensuring all required libraries are available during serving.

::

    my_project/
    |── train.py
    └── custom_package.whl

Then the following code can log your model with the custom package:

.. code-block:: python
    :caption: train.py

    import mlflow
    from custom_package import my_func

    class MyModel(mlflow.pyfunc.PythonModel):
        def predict(self, context, model_input):
            x = my_func(model_input)
            # .. your prediction logic
            return prediction

    # Log the model
    with mlflow.start_run() as run:
        mlflow.pyfunc.log_model(
            python_model=MyModel(),
            artifact_path="model",

>             extra_pip_requirements=["code/custom_package.whl"],

Reviewer: I think this is not the recommended way.

I recommend to upload the private package wheel file to an internal server or clould storage with a downloading link,
then when logging model, log extra_pip_requirements=[<wheel_link>]

Reviewer 2: Within the note block showing an example object store path for 'once you're ready for production' is ideal, while the example you have here is perfect for dev testing while building a solution.

Author: Thank you for the comments both, I'll add the disclaimer.

Author: @WeichenXu123 @BenWilson2 I've added the note section, can you take another look?

Author: <img width="1196" alt="image" src="https://github.com/user-attachments/assets/c8f8ab7d-05e6-45cd-9704-591941b86ffa">

When MLflow loads the model for serving, the ``code`` directory will be added to the system path so that you can use the module in your model
code like ``from utils import my_func``. You can also specify a directory path as ``code_paths`` to save multiple files under the directory:

Use of ``code_paths`` Option for a Custom Library
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To include custom libraries or libraries from a private mirror server when logging your model, the ``code_paths`` option can be used. 

Reviewer 3: ```suggestion
To include a custom library that is not publicly available on PyPI when logging your model, the ``code_paths`` argument can be used. 
```

> This option allows you to upload .whl files or other dependencies alongside your model, ensuring all required libraries are available during serving.

Reviewer 3: ```suggestion
This option allows you to upload ``.whl`` files or other dependencies alongside your model, ensuring all required libraries are available during serving.
```

> .. note::

Reviewer 3: ```suggestion
.. warning::
```

    The following example demonstrates a quick method for including custom libraries for development purposes. 
    This approach is not recommended for production environments. 
    For production usage, upload libraries to a custom PyPI server or a cloud storage to ensure reliable and secure access.

> 
> ::
> 
>     my_project/
>     |── train.py
>     └── custom_package.whl

Reviewer 3: ```suggestion

<what this is>

::

    my_project/
    |── train.py
    └── custom_package.whl
```

Can we explain what this is?

Then the following code can log your model with the custom package:

.. code-block:: python
    :caption: train.py

    import mlflow
    from custom_package import my_func

    class MyModel(mlflow.pyfunc.PythonModel):
        def predict(self, context, model_input):
            x = my_func(model_input)
            # .. your prediction logic
            return prediction

    # Log the model
    with mlflow.start_run() as run:
        mlflow.pyfunc.log_model(
            python_model=MyModel(),
            artifact_path="model",
            extra_pip_requirements=["code/custom_package.whl"],
            input_example=input_data,

>             code_paths=["custom_package.whl"],

Reviewer 3: Does the `code_path` accept wheels? I thought it only accepts `.py` files or directories.

Reviewer 3: cc @WeichenXu123 to confirm.

Author: I conducted an experiment on my local with the following code and it works ok, so i guess this usecase is legit.

```python
import mlflow
from mlflow.tracking import MlflowClient
import numpy as np

# from wheel
import sympy

class MyModel(mlflow.pyfunc.PythonModel):
    def predict(self, context, model_input):
        x = [float(sympy.acos(i)) for i in model_input]
        return x

client = MlflowClient()
mlflow.set_tracking_uri("http://127.0.0.1:5000")
experiment_name = "custom_library"
experiment = client.get_experiment_by_name(experiment_name)
if experiment is None:
    experiment_id = client.create_experiment(experiment_name)
else:
    experiment_id = experiment.experiment_id
with mlflow.start_run(experiment_id=experiment_id) as run:
    model_info = mlflow.pyfunc.log_model(
        python_model=MyModel(),
        artifact_path="model",
        input_example=np.array([1.0, 0.0]),
        pip_requirements=["code/sympy-1.13.3-py3-none-any.whl"],
        code_paths=["sympy-1.13.3-py3-none-any.whl"],
    )
loaded_model = mlflow.pyfunc.load_model(model_uri=model_info.model_uri)
print(loaded_model.predict(np.array([1.0, 0.0])))
```

Reviewer 3: Got it. Let's update the `code_paths` docstring. It doesn't sound like it accepts arbitrary files. If we're missing tests, we should also add tests to ensure the example in this PR works.

Author: Sure, let me work on the doc update and potential unit tests in a separate PR!
